Assets/Scripts/ServiceManagers/StatsManager.cs

```diff
@@ -0,0 +1,249 @@
+        {
+            _playerStats[productUserId] = await QueryPlayerStatsAsync(productUserId);
+
+            // Because statistics can change achievements, refresh the 
+            // achievements service as well.
+            EOSAchievementManager.Instance.Refresh();
+        }
+
+        /// <summary>
+        /// Ingest a given stat for the current player.
+        /// </summary>
+        /// <param name="statName">The stat name.</param>
+        /// <param name="ingestAmount">
+        /// The amount to "ingest" for the stat.
+        /// </param>
+        /// <returns>
+        /// A task so that this function can be awaitable.
+        /// </returns>
+        public Task IngestStatAsync(string statName, int ingestAmount)
+        {
+            ProductUserId userId = EOSManager.Instance.GetProductUserId();
+            IngestStatOptions ingestOptions = new()
+            {
+                LocalUserId = userId,
+                TargetUserId = userId,
+                Stats = new[] { new IngestData() { StatName = statName, IngestAmount = ingestAmount } }
+            };
+
+            TaskCompletionSource<object> taskCompletionSource = new();
+
+            GetEOSStatsInterface().IngestStat(ref ingestOptions, null, (ref IngestStatCompleteCallbackInfo data) =>
+            {
+                taskCompletionSource.SetResult(null);
+                EOSAchievementManager.Instance.Refresh();
+            });
+
+            return taskCompletionSource.Task;
+        }
+
+        /// <summary>
+        /// Queries from the server the stats pertaining to the user associated
+        /// to the given ProductUserId.
+        /// </summary>
+        /// <param name="productUserId">
+        /// The ProductUserId associated with the player to get the statistics
+        /// for.
+        /// </param>
+        /// <returns>
+        /// List of stats for the indicated player.
+        /// </returns>
+        private static Task<List<Stat>> QueryPlayerStatsAsync(ProductUserId productUserId)
+        {
+            if (!productUserId.IsValid())
+            {
+                Log("Invalid product user id sent in!");
+                return Task.FromResult(new List<Stat>());
+            }
+            var statInterface = GetEOSStatsInterface();
+
+            QueryStatsOptions statsOptions = new()
+            {
+                LocalUserId = productUserId,
+                TargetUserId = productUserId
+            };
+
+            TaskCompletionSource<List<Stat>> tcs = new();
+
+            statInterface.QueryStats(ref statsOptions, null, (ref OnQueryStatsCompleteCallbackInfo queryStatsCompleteCallbackInfo) =>
+            {
+                if (queryStatsCompleteCallbackInfo.ResultCode != Result.Success)
+                {
+                    // TODO: handle error
+                    Log($"Failed to query stats, result code: {queryStatsCompleteCallbackInfo.ResultCode}");
+                    tcs.SetResult(new List<Stat>());
```

Comment:
Do you prefer that to the setting exception done in some other places?

Reply:
Good call - so I updated it to _not_ call the `SetResult` function, but instead call the `SetException` function passing an exception with a message equal to what was previously sent to the `Log` function.